src/object-pool.ts

```diff
@@ -1,10 +1,287 @@
+			}
+		}
+
+		this.state.autoIncrease = result.autoIncrease;
+		this.state.maxSize = result.maxSize;
+		this.state.increaseFactor = result.increaseFactor;
+		this.state.increaseBreakPoint = result.increaseBreakPoint;
+
+		this.startSize = result.objectCount;
+	}
+
+	public utilization(n: number = 0): number {
+		if (!this.isValidState(this.state)) {
+			return NaN;
+		}
+		if (this.state.objectCount === 0) {
+			return Infinity;
+		}
+
+		let num: number = n;
+		if (isNaN(num)) {
+			num = 0;
+		}
+
+		const freeObj = this.state.elements.length - num;
+		return (this.state.objectCount - freeObj) / this.state.objectCount;
+	}
+	public isAboveThreshold(n: number = 0): boolean {
+		return this.utilization(n) >= this.state.increaseBreakPoint;
+	}
+
+	public get(): T | null {
+		if (!this.isValidState(this.state)) {
+			return null;
+		}
+
+		if (this.state.autoIncrease && this.isAboveThreshold(1)) {
+			this.increase(Math.ceil(this.state.objectCount * this.state.increaseFactor));
+		}
+
+		let result = this.state.elements.pop();
+
+		if (result === undefined) {
+			return null;
+		} else {
+			return result;
+		}
+	}
+	public allocate(n: number = 1): Array<T> {
+		if (!this.isValidState(this.state)) {
+			return [];
+		}
+
+		let num: number;
+		if (!this.isInteger(n) || n < 1) {
+			num = 1;
+		} else {
+			num = n;
+		}
+
+		let result: Array<T> = [];
+
+		while (this.state.autoIncrease && this.isAboveThreshold(num)) {
+			this.increase(Math.ceil(this.state.objectCount * this.state.increaseFactor));
+		}
+
+		for (let i = 0; i < num && this.state.elements.length; i++) {
+			result.push(this.state.elements.pop()!);
+		}
+
+		return result;
+	}
+	public increase(n: number): void {
+		if (!this.isValidState(this.state)) {
+			return;
+		}
+		if (!this.isInteger(n)) {
+			return;
+		}
+
+		for (let i = 0; i < n && this.state.objectCount < this.state.maxSize; i++) {
+			this.store(new this.poolObj());
+			this.state.objectCount++;
+		}
+	}
+
+	public release(object: T): void {
+		if (!this.poolObj || !this.poolObj.cleanObj) {
+			return;
+		}
+
+		this.poolObj.cleanObj(object);
+		this.store(object);
+	}
+	public store(object: T): void {
+		if (!this.isValidState(this.state)) {
+			return;
+		}
+
+		this.state.elements.push(object);
+	}
+
+	public isInteger(n: number): boolean {
+		if (typeof n !== 'number') {
+			return false;
+		}
+		if (n % 1 !== 0) {
+			return false;
+		}
+
+		return true;
+	}
+	public isValidState(state: ArmorObjectPoolState<T>): boolean {
+		if (!state) {
+			return false;
+		}
+		if (state.type !== 'opState') {
+			return false;
+		}
+
+		if (!Array.isArray(state.elements)) {
+			return false;
+		}
+
+		if (!this.isInteger(state.objectCount) || state.objectCount < 0) {
+			return false;
+		}
+		if (!this.isInteger(state.maxSize) || state.maxSize < 1) {
+			return false;
+		}
+
+		if (typeof state.increaseFactor !== 'number' || !(state.increaseFactor >= 0)) {
```

Comment:
For `!(state.increaseFactor >= 0)` it's a little more clear to invert the statement and say `|| state.increaseFactor < 0`, if I understand what's happening here.

Reply:
I used !(state.increaseFactor >= 0) because it handles the NaN case. I added an isFloat function to make the if more clear.